File: source/dotnet/source/cmf.bus/cmf.eventing.berico/OutboundHeadersProcessor.cs
using System;
using System.Collections.Generic;
using System.DirectoryServices.AccountManagement;
using System.Linq;
using System.Text;

using cmf.bus;

namespace cmf.eventing.berico
{
    public class OutboundHeadersProcessor : IOutboundEventProcessor
    {
        public virtual void ProcessOutbound(ref object ev, ref bus.Envelope env, IDictionary<string, object> context)
        {
            Guid messageId = env.GetMessageId();
            messageId = Guid.Equals(Guid.Empty, messageId) ? Guid.NewGuid() : messageId;
            env.SetMessageId(messageId.ToString());

            Guid correlationId = env.GetCorrelationId();

            string messageType = env.GetMessageType();
            messageType = string.IsNullOrEmpty(messageType) ? this.GetMessageType(ev) : messageType;
            env.SetMessageType(messageType);

            string messageTopic = env.GetMessageTopic();
            messageTopic = string.IsNullOrEmpty(messageTopic) ? this.GetMessageTopic(ev) : messageTopic;
            if (Guid.Empty != correlationId)
            {
                messageTopic = messageTopic + "#" + correlationId.ToString();
            }
            env.SetMessageTopic(messageTopic);

            string senderIdentity = env.GetSenderIdentity();
            senderIdentity = string.IsNullOrEmpty(senderIdentity) ? UserPrincipal.Current.DistinguishedName.Replace(",", ", ") : senderIdentity;
            env.SetSenderIdentity(senderIdentity);
        }

        public string GetMessageTopic(object ev)
        {
            string topic = ev.GetType().FullName;

            try
            {
                object[] attributes = ev.GetType().GetCustomAttributes(typeof(EventAttribute), true);
                EventAttribute attr = attributes.OfType<EventAttribute>().FirstOrDefault();
                if (null != attr)
                {
                    topic = attr.EventTopic;
                }
            }
            catch { }

            return topic;
        }

        public string GetMessageType(object ev)
        {
            string type = ev.GetType().FullName;

            try
            {
                object[] attributes = ev.GetType().GetCustomAttributes(typeof(EventAttribute), true);
                EventAttribute attr = attributes.OfType<EventAttribute>().FirstOrDefault();
                if (null != attr)
                {
                    type = attr.EventType;
                }
            }
            catch { }

            return type;
        }
    }
}

File: source/java/src/main/java/cmf/eventing/berico/OutboundHeadersProcessor.java
package cmf.eventing.berico;

import java.util.UUID;

import org.apache.commons.lang.StringUtils;

import cmf.bus.berico.EnvelopeHelper;
import cmf.eventing.Event;
import cmf.security.IUserInfoRepository;

public class OutboundHeadersProcessor implements IOutboundEventProcessor {

	protected IUserInfoRepository userInfoRepo;
	
	
	public OutboundHeadersProcessor(IUserInfoRepository userInfoRepository) {
		this.userInfoRepo = userInfoRepository;
	}
	
	
	@Override
	public void processOutbound(ProcessingContext context) throws Exception {
		
		EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());
		
        UUID messageId = env.getMessageId();
        messageId = (null == messageId) ? UUID.randomUUID() : messageId;
        env.setMessageId(messageId);

        UUID correlationId = env.getCorrelationId();

        String messageType = env.getMessageType();
        messageType = StringUtils.isBlank(messageType) ? this.getMessageType(context.getEvent()) : messageType;
        env.setMessageType(messageType);

        String messageTopic = env.getMessageTopic();
        messageTopic = StringUtils.isBlank(messageTopic) ? this.getMessageTopic(context.getEvent()) : messageTopic;
        if (null != correlationId)
        {
            messageTopic = messageTopic + "#" + correlationId.toString();
        }
        env.setMessageTopic(messageTopic);
        
        String senderIdentity = env.getSenderIdentity();
        senderIdentity = StringUtils.isBlank(senderIdentity) ? userInfoRepo.getDistinguishedName(System.getProperty("user.name")).replaceAll(",", ", ") : senderIdentity;
        env.setSenderIdentity(senderIdentity);
	}
	
	
	protected String getMessageType(Object event) {
		
		String messageType = event.getClass().getCanonicalName();
		
		Event eventMetadata = event.getClass().getAnnotation(Event.class);
		
		if (null != eventMetadata) {
			messageType = eventMetadata.type();
		}
		
		return messageType;
	}
	
	protected String getMessageTopic(Object event) {
		
		String messageTopic = event.getClass().getCanonicalName();
		
		Event eventMetadata = event.getClass().getAnnotation(Event.class);
		
		if (null != eventMetadata) {
			messageTopic = eventMetadata.topic();
		}
		
		return messageTopic;
	}
}
